source/MRMesh/MRMeshLoadObj.cpp

```diff
@@ -22,861 +24,1035 @@
+                currentMaterial = {};
             }
-            if ( !parseError.empty() )
-                return unexpected( parseError );
+            currentMaterialName = line.substr( 6, std::string_view::npos );
+            boost::trim( currentMaterialName );
         }
+        break;
+        case MtlElement::DiffuseColor:
+        {
+            const auto li = group.end - 1;
+            std::string_view line( data + newlines[li], newlines[li + 1] - newlines[li + 0] );
 
-        if ( !currentMaterialName.empty() )
-            result.emplace( std::move( currentMaterialName ), std::move( currentMaterial ) );
-        return result;
+            auto res = parseMtlColor( line, currentMaterial.diffuseColor );
+            if ( !res.has_value() )
+                parseError = std::move( res.error() );
+        }
+        break;
+        case MtlElement::DiffuseTexture:
+        {
+            const auto li = group.end - 1;
+            std::string_view line( data + newlines[li], newlines[li + 1] - newlines[li + 0] );
+
+            auto res = parseMtlTexture( line, currentMaterial.diffuseTextureFile );
+            if ( !res.has_value() )
+                parseError = std::move( res.error() );
+        }
+        break;
+        default:
+            break;
+        }
+        if ( !parseError.empty() )
+            return unexpected( parseError );
     }
+
+    if ( !currentMaterialName.empty() )
+        result.emplace( std::move( currentMaterialName ), std::move( currentMaterial ) );
+    return result;
 }
 
-namespace MR
+struct MaterialScope
 {
-
-namespace MeshLoad
+    size_t fId; // material begins with this face
+    std::string mtName;
+};
+struct ObjectScope
 {
-
-Expected<std::vector<NamedMesh>> fromSceneObjFile( const std::filesystem::path& file, bool combineAllObjects,
-                                                                const ObjLoadSettings& settings /*= {}*/ )
+    size_t fId; // object begins with this face
+    std::string objName;
+};
+
+Expected<MeshLoad::NamedMesh> loadSingleModelFormObj(
+    const std::filesystem::path& dir,
+    const Vector<Vector3d, VertId>& points,  // all points from file
+    const std::vector<Color>& colors,     // all colors from file
+    const std::vector<UVCoord>& uvCoords, // all uvs from file
+    const std::vector<ObjFace>& faces,    // all faces from file
+    const std::vector<MaterialScope>& materialScope, // all material scopes from file
+    size_t minFace, size_t maxFace,       // this model faces span in `faces`, max face excluding
+    const MeshLoad::ObjLoadSettings& settings,
+    const MtlLibrary* mtl ) // optional materials, if nullptr `materialScope` will be ignored
 {
-    std::ifstream in( file, std::ios::binary );
-    if ( !in )
-        return unexpected( std::string( "Cannot open file for reading " ) + utf8string( file ) );
+    MR_TIMER;
+    using ParallelIndicesMap = ParallelHashMap<Vector2i, VertId>;
```

Comment:
Replace `Vector2i` with a new struct. `repr.x` and `repr.y` are non-obvious names.